# **Advanced Data Science for Innovation - Assignment 1**


## NBA Career Prediction: Predict 5-Year Longevity for NBA Rookies
**Student Name:** Sean Williams

**Team Name:** Group 1
* Nuwan Munasinghe
* Wenying Wu
* Nathan Fragar
* Sean Williams
* Carol Myhill

## Environment Setup

In [1]:
import pandas as pd
from importlib.machinery import SourceFileLoader
dataexp = SourceFileLoader('explore', '../src/data/explore.py').load_module()

# 1. Data Exploration

**<u>Collect initial data</u>**

In [2]:
df_train = pd.read_csv('../data/raw/train.csv')
df_test = pd.read_csv('../data/raw/test.csv')

---
**<u>Display train dateset information</u>**

In [3]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Id           8000 non-null   int64  
 1   GP           8000 non-null   int64  
 2   MIN          8000 non-null   float64
 3   PTS          8000 non-null   float64
 4   FGM          8000 non-null   float64
 5   FGA          8000 non-null   float64
 6   FG%          8000 non-null   float64
 7   3P Made      8000 non-null   float64
 8   3PA          8000 non-null   float64
 9   3P%          8000 non-null   float64
 10  FTM          8000 non-null   float64
 11  FTA          8000 non-null   float64
 12  FT%          8000 non-null   float64
 13  OREB         8000 non-null   float64
 14  DREB         8000 non-null   float64
 15  REB          8000 non-null   float64
 16  AST          8000 non-null   float64
 17  STL          8000 non-null   float64
 18  BLK          8000 non-null   float64
 19  TOV   

**Observations:**
* The training dataset contains 8000 observations
* The target (TARGET_5Yrs) is an Integer with no missing values
* There are 20 features within the dataset (2 Integer and 18 Float)
* There are no categorical features within the dataset
* There are no null values across all faetures of the dataset
---
**<u>Display test dateset information</u>**

In [4]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3799 entries, 0 to 3798
Data columns (total 20 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Id       3799 non-null   int64  
 1   GP       3799 non-null   int64  
 2   MIN      3799 non-null   float64
 3   PTS      3799 non-null   float64
 4   FGM      3799 non-null   float64
 5   FGA      3799 non-null   float64
 6   FG%      3799 non-null   float64
 7   3P Made  3799 non-null   float64
 8   3PA      3799 non-null   float64
 9   3P%      3799 non-null   float64
 10  FTM      3799 non-null   float64
 11  FTA      3799 non-null   float64
 12  FT%      3799 non-null   float64
 13  OREB     3799 non-null   float64
 14  DREB     3799 non-null   float64
 15  REB      3799 non-null   float64
 16  AST      3799 non-null   float64
 17  STL      3799 non-null   float64
 18  BLK      3799 non-null   float64
 19  TOV      3799 non-null   float64
dtypes: float64(18), int64(2)
memory usage: 593.7 KB


**Observations:**
* The test dataset contains 3799 observations
* There is no target column
* There are 20 features within the dataset (2 Integer and 18 Float).
* Names and datatypes of features align with features availble in the training dataset
* No categorical features within the dataset
* There are no null values across all features of the dataset
---
**<u>Display first 20 rows of train dataset</u>**

In [5]:
df_train.head(20)

Id  GP   MIN   PTS  FGM   FGA   FG%  3P Made  3PA   3P%  ...  FTA  \
0   10556  80  24.3   7.8  3.0   6.4  45.7      0.1  0.3  22.6  ...  2.9   
1    5342  75  21.8  10.5  4.2   7.9  55.1     -0.3 -1.0  34.9  ...  3.6   
2    5716  85  19.1   4.5  1.9   4.5  42.8      0.4  1.2  34.3  ...  0.6   
3   13790  63  19.1   8.2  3.5   6.7  52.5      0.3  0.8  23.7  ...  1.5   
4    5470  63  17.8   3.7  1.7   3.4  50.8      0.5  1.4  13.7  ...  0.5   
5    1961  88  20.0   8.8  3.7   8.7  43.6     -0.2 -0.5   6.9  ...  2.8   
6    2677  70  20.6   7.0  3.2   5.4  58.9      0.1 -0.1  -3.6  ...  1.4   
7   11722  57  17.6   5.4  1.7   4.7  37.2      0.9  2.8  33.2  ...  1.8   
8    6148  46  26.5   9.3  3.1   7.2  44.0      0.6  1.7  27.5  ...  2.3   
9    3040  64  33.4  16.9  6.4  11.3  59.0      0.3  1.1 -13.0  ...  4.9   
10   4212  72  30.6  13.2  4.9  11.0  46.9      1.0  2.5  48.1  ...  2.8   
11   5923  47  15.2   6.5  2.5   6.5  38.3      0.7  2.5  39.7  ...  1.1   
12  13144  41   9.8   2.3  0.9   2.3  37.9     -0.4 -0.6  -4.7  ...  1.3   
13   8884  64  29.7  12.0  4.2   9.6  45.2      0.8  2.1  28.3  ...  3.3   
14  12443  90  31.1  12.9  4.9  10.3  47.0      0.3  1.1  17.0  ...  2.6   
15   2874  45   5.2   1.3  0.5   1.4  41.0      0.1  0.1  21.1  ...  0.2   
16   8040  29   7.5   1.9  0.7   2.6  30.1      0.3  1.0  26.8  ...  0.4   
17    762  53  13.8   6.7  2.1   4.7  43.5      0.2  0.7  19.8  ...  3.8   
18   3062  83   8.5   2.2  0.9   1.8  48.4     -0.3 -1.0   5.6  ...  0.9   
19    829  64  10.4   5.1  2.2   4.9  45.5     -0.1 -0.2  -6.1  ...  1.8   

     FT%  OREB  DREB   REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0   72.1   2.2   2.0   3.8  3.2  1.1  0.2  1.6            1  
1   67.8   3.6   3.7   6.6  0.7  0.5  0.6  1.4            1  
2   75.7   0.6   1.8   2.4  0.8  0.4  0.2  0.6            1  
3   66.9   0.8   2.0   3.0  1.8  0.4  0.1  1.9            1  
4   54.0   2.4   2.7   4.9  0.4  0.4  0.6  0.7            1  
5   75.1   1.8   2.9   4.7  1.8  0.4  0.3  1.1            1  
6   56.7   2.9   4.6   7.6  0.6  0.4  0.7  1.2            1  
7   78.3   0.8   1.7   2.6  0.4  0.6  0.6  0.3            1  
8   82.1   1.3   2.8   4.1  1.8  0.7  0.6  1.7            1  
9   74.9   3.4   8.0  11.9  0.8  0.4  2.5  2.4            1  
10  74.1   1.0   3.1   3.9  4.0  1.1  0.2  1.5            1  
11  75.5   0.6   1.3   2.0  0.3  0.4  0.3  0.7            1  
12  64.7   1.5   1.9   3.5  0.2  0.3  0.4  0.4            0  
13  76.8   1.5   3.1   4.8  3.8  1.5 -0.0  2.0            1  
14  93.4   1.6   3.4   4.7  2.3  0.6  0.4  2.5            1  
15  73.3   0.3   0.7   1.0  0.2  0.2  0.2  0.5            1  
16  68.0   0.4   0.8   1.2  0.5  0.3  0.1  0.7            1  
17  75.1   1.1   1.9   3.1  0.2  0.4  0.6  0.7            1  
18  63.4   0.2   0.6   0.8  1.3  0.4  0.2  0.8            1  
19  43.6   0.7   1.0   1.6  1.2  0.5  0.1  1.2            1  

[20 rows x 21 columns]

**Observations:**
* Negative values noted in 3P Made, 3PA and 3P% features
---
**<u>Display first 20 rows of test dataset</u>**

In [6]:
df_test.head(20)

Id  GP   MIN   PTS   FGM   FGA   FG%  3P Made  3PA   3P%  FTM  FTA  \
0      1  56   9.1   4.0   1.6   3.7  43.7      0.1  0.3   7.3  0.7  1.2   
1   8194  43  19.3  10.1   3.7   8.1  46.0      0.6  1.7  35.1  1.8  2.5   
2      3  82  33.9  11.3   4.9  10.6  45.6      0.5  1.9  44.8  1.8  2.7   
3   8196  86  44.7  18.8   6.8  15.9  42.9      0.5  1.8  13.5  4.5  6.3   
4   8197  58  12.3   4.7   1.6   4.0  40.0      0.5  1.7  38.7  1.1  1.3   
5      6  59  16.1   7.1   2.8   4.5  62.0     -0.1 -0.6   3.4  1.7  2.7   
6   8198  45   8.3   3.2   1.4   2.8  46.4     -0.5 -1.6   8.9  0.5  0.7   
7   8202  53  12.1   4.7   1.8   4.1  44.1      0.1  0.2  25.7  0.9  1.0   
8     14  41  21.6   7.9   3.1   6.0  49.6      0.3  0.9 -14.8  1.8  2.2   
9   8207  25  10.0   2.6   1.0   2.3  45.0      0.2  0.7  11.7  0.4  0.5   
10    17  72  12.6   3.5   1.3   3.2  41.9      0.1  0.1  14.9  0.5  0.6   
11  8214  67  21.3  11.1   4.6   8.3  53.9      0.0  0.2  22.5  1.9  2.4   
12    25  47  31.1  11.5   4.1   9.9  42.1      1.0  3.0  28.7  2.1  2.6   
13  8217  87  20.7  10.9   4.5   9.8  48.4      0.5  1.1  28.1  1.4  2.3   
14  8219  82  43.4  17.3   7.0  15.4  46.2      1.0  2.5  31.2  2.4  2.9   
15  8220  68  42.4  23.1  10.0  20.8  48.7      0.6  1.7   7.0  3.4  5.1   
16  8228  78  23.9   8.1   3.1   5.8  52.1     -0.1 -0.1  33.8  2.1  3.2   
17    38  52  21.2   9.7   3.8   8.0  49.0      0.7  2.1  35.6  1.6  2.2   
18    39  27   8.5   1.6   0.7   1.4  50.4     -0.2 -0.6  22.4  0.1  0.2   
19  8230  77  22.2   6.9   3.0   6.6  45.1      0.1  0.2  -8.2  1.0  1.6   

     FT%  OREB  DREB  REB  AST  STL  BLK  TOV  
0   63.4   1.2   0.8  1.7  0.4  0.2  0.3  0.8  
1   75.3   0.5   0.9  1.5  3.5  0.6 -0.0  1.8  
2   71.2   1.3   3.3  4.5  2.5  1.3  0.3  2.0  
3   70.9   1.5   3.2  5.0  4.1  0.9  0.1  3.6  
4   76.9   0.2   0.6  0.9  1.5  0.5 -0.4  0.9  
5   58.6   1.8   2.6  4.6  0.6  0.6  0.4  0.9  
6   69.3   0.5   0.5  1.2  0.2  0.1  0.2  0.4  
7   80.2   0.6   1.1  1.6  0.8  0.4 -0.2  0.6  
8   77.6   2.4   2.6  5.1  1.4  0.5  0.4  1.4  
9   84.1   0.4   1.5  1.8  0.4  0.2  0.2  0.5  
10  74.8   0.5   1.4  1.8  1.1  0.4  0.2  0.8  
11  81.3   1.6   2.4  4.0  3.3  1.3  0.3  1.7  
12  72.3   1.0   3.3  4.4  1.9  0.9  0.2  1.2  
13  61.2   0.9   1.4  2.2  3.2  1.0  0.0  2.3  
14  84.3   2.1   3.5  5.3  4.3  1.4  0.2  3.0  
15  69.4   2.0   4.3  6.4  4.2  1.7  0.3  2.3  
16  64.3   1.4   2.3  3.7  1.4  0.6  0.3  1.8  
17  67.8   1.0   2.5  3.4  1.7  0.9  0.3  1.5  
18  55.8   0.6   1.1  1.5  0.4  0.3  0.0  0.4  
19  49.2   2.2   2.3  4.1  1.4  0.4  0.6  1.5

**Observations:**
* Negative values noted in 3P Made, 3PA, 3P% and BLK features
---
**<u>Display statistical information for all numeric features of train dataset</u>**

In [7]:
df_train.describe()

Id           GP          MIN          PTS          FGM  \
count   8000.000000  8000.000000  8000.000000  8000.000000  8000.000000   
mean    6856.971000    62.777875    18.576662     7.267088     2.807037   
std     3977.447579    17.118774     8.935263     4.318732     1.693373   
min        4.000000    -8.000000     2.900000     0.800000     0.300000   
25%     3413.750000    51.000000    12.000000     4.100000     1.600000   
50%     6787.500000    63.000000    16.800000     6.300000     2.400000   
75%    10299.250000    74.000000    23.500000     9.500000     3.700000   
max    13798.000000   123.000000    73.800000    34.200000    13.100000   

               FGA          FG%      3P Made          3PA          3P%  ...  \
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000  ...   
mean      6.231212    44.608900     0.264525     0.816562    19.583700  ...   
std       3.584559     6.155453     0.384093     1.060964    16.003155  ...   
min       0.800000    21.300000    -1.100000    -3.100000   -38.500000  ...   
25%       3.600000    40.400000     0.000000     0.100000     8.400000  ...   
50%       5.400000    44.400000     0.300000     0.800000    19.500000  ...   
75%       8.100000    48.700000     0.500000     1.500000    30.600000  ...   
max      28.900000    67.200000     1.700000     4.700000    82.100000  ...   

               FTA          FT%         OREB         DREB          REB  \
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000   
mean      1.947788    71.365825     1.077838     2.168500     3.245300   
std       1.252352    10.430447     0.785670     1.392224     2.085154   
min       0.000000   -13.300000     0.000000     0.200000     0.300000   
25%       1.000000    65.000000     0.500000     1.100000     1.700000   
50%       1.700000    71.400000     0.900000     1.900000     2.800000   
75%       2.600000    77.500000     1.500000     2.900000     4.300000   
max      11.100000   168.900000     5.500000    11.000000    15.900000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  8000.000000  8000.000000  8000.000000  8000.000000  8000.000000  
mean      1.624513     0.648687     0.245212     1.257763     0.833625  
std       1.355986     0.407626     0.821037     0.723270     0.372440  
min       0.000000     0.000000   -17.900000     0.100000     0.000000  
25%       0.700000     0.300000     0.100000     0.700000     1.000000  
50%       1.300000     0.600000     0.200000     1.100000     1.000000  
75%       2.200000     0.900000     0.400000     1.600000     1.000000  
max      12.800000     3.600000    18.900000     5.300000     1.000000  

[8 rows x 21 columns]

**Observations:**
* Minimum value confirms negative values noted in GP, 3P Made, 3PA, 3P%, FT% and BLK features
---
**<u>Display statistical information for all numeric features of test dataset</u>**

In [8]:
df_test.describe()

Id           GP          MIN          PTS          FGM  \
count   3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean    7010.614109    62.853909    18.650224     7.328034     2.835404   
std     3954.173641    17.151740     8.727259     4.294724     1.688427   
min        1.000000     6.000000     3.700000     0.700000     0.300000   
25%     3644.000000    51.000000    12.200000     4.200000     1.600000   
50%     7062.000000    63.000000    17.000000     6.400000     2.500000   
75%    10402.500000    74.000000    23.300000     9.400000     3.700000   
max    13792.000000   126.000000    68.000000    33.000000    13.400000   

               FGA          FG%      3P Made          3PA          3P%  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      6.302580    44.599079     0.255962     0.796920    19.234746   
std       3.579221     6.040168     0.380987     1.052862    15.968989   
min       0.800000    25.100000    -1.000000    -2.700000   -38.000000   
25%       3.700000    40.500000    -0.000000     0.100000     8.500000   
50%       5.500000    44.600000     0.300000     0.800000    19.400000   
75%       8.100000    48.500000     0.500000     1.500000    30.250000   
max      26.200000    74.600000     1.600000     4.300000    73.800000   

               FTM          FTA          FT%         OREB         DREB  \
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000   
mean      1.399842     1.953567    71.612924     1.096025     2.179495   
std       0.926140     1.250376    10.457336     0.785678     1.371935   
min       0.000000     0.000000    23.700000     0.000000     0.200000   
25%       0.700000     1.000000    65.000000     0.500000     1.200000   
50%       1.200000     1.700000    71.500000     0.900000     1.900000   
75%       1.900000     2.600000    78.000000     1.500000     2.900000   
max       7.800000     9.800000   127.100000     6.900000    12.000000   

               REB          AST          STL          BLK          TOV  
count  3799.000000  3799.000000  3799.000000  3799.000000  3799.000000  
mean      3.275783     1.636483     0.653593     0.257726     1.257910  
std       2.070646     1.335496     0.410573     0.639660     0.712449  
min       0.300000     0.000000     0.000000    -7.100000     0.100000  
25%       1.800000     0.600000     0.400000     0.100000     0.700000  
50%       2.800000     1.300000     0.600000     0.200000     1.100000  
75%       4.300000     2.300000     0.900000     0.400000     1.600000  
max      18.500000     9.000000     2.700000    14.800000     5.200000

**Observations:**
* Minimum value confirms negative values noted in 3P Made, 3PA, 3P% and BLK features
---
**<u>Display details of features with missing data in train dataset</u>**

In [9]:
dataexp.print_na_info(df_train)

Empty DataFrame
Columns: [feature, null_value_count, null_value_percent]
Index: []

**Observations:**
* No missing values in train data set
---
**<u>Display details of features with missing data in test dataset</u>**

In [10]:
dataexp.print_na_info(df_test)

Empty DataFrame
Columns: [feature, null_value_count, null_value_percent]
Index: []

**Observations:**
* No missing values in test data set
---
**<u>Check for any duplicate data in train dataset</u>**

In [11]:
df_train[df_train.duplicated()]

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV, TARGET_5Yrs]
Index: []

[0 rows x 21 columns]

In [12]:
df_train[df_train.duplicated(['Id'])]

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV, TARGET_5Yrs]
Index: []

[0 rows x 21 columns]

**Observations:**
* No duplicates data in train dataset
---
**<u>Check for any duplicate data in test dataset</u>**

In [13]:
df_test[df_test.duplicated()]

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV]
Index: []

In [14]:
df_test[df_test.duplicated(['Id'])]

Empty DataFrame
Columns: [Id, GP, MIN, PTS, FGM, FGA, FG%, 3P Made, 3PA, 3P%, FTM, FTA, FT%, OREB, DREB, REB, AST, STL, BLK, TOV]
Index: []

**Observations:**
* No duplicate data in test dataset
---
**<u>Determine number of observations with negative values in train dataset</u>**

In [15]:
dataexp.print_negative_info(df_train)

feature  negative_value_count  negative_value_percent
0      3PA                  1658                 20.7250
1  3P Made                  1629                 20.3625
2      BLK                  1048                 13.1000
3      3P%                   878                 10.9750
4       GP                     2                  0.0250
5      FT%                     1                  0.0125

**Observations:**
* High occurence of observations with negative 3PA, 3P Made, 3P% and BLK values. These basketball stats cannot be negative
* GP negatives counts are low. Possible data issue. Drop rows from data set
* FT% negative counts are low. Possible data issue. Drop rows from data set
---
**<u>Determine number of observations with negative values in test dataset</u>**

In [16]:
dataexp.print_negative_info(df_test)

feature  negative_value_count  negative_value_percent
0  3P Made                   775               20.400105
1      3PA                   773               20.347460
2      BLK                   456               12.003159
3      3P%                   435               11.450382

**Observations:**
* High occurence of observations with negative 3P values and BLK values. 3 Point could be an important statistic in determining player longetivity so dropping column might not be desirable
* No GP or FT% negative values.
---
**<u>Print frequency of Target variable</u>**

In [17]:
print('Frequency of target variable in original dataset')
df_train['TARGET_5Yrs'].value_counts(normalize=True)

Frequency of target variable in original dataset


1    0.833625
0    0.166375
Name: TARGET_5Yrs, dtype: float64

**Observations:**
* Data is heavily skewed to Positive outcome (i.e. most players have NBA careers lasting 5 years or more)
---
**<u>Data Preparation Tasks</u>**
* Drop rows where GP < 0. This will  only affect training data set. Test data set needs to have 3799 observations
* Drop rows where FT% < 0. This will  only affect training data set. Test data set needs to have 3799 observations
* Construct new features
* Drop ID column. Column identifies a player and has no relevance to modelling
* Scale all remaining features